waiter/src/waiter/interstitial.clj

```diff
@@ -0,0 +1,238 @@
+                      current-state)
+
+                    scheduler-state-chan
+                    (let [scheduler-messages message
+                          available-service-ids' (process-scheduler-messages
+                                                   interstitial-state-atom service-id->service-description
+                                                   available-service-ids scheduler-messages)]
+                      (when-not (interstitial-state-initialized? interstitial-state-atom)
+                        (log/info "interstitial state has been initialized with services from the scheduler")
+                        (swap! interstitial-state-atom assoc :initialized? true))
+                      (assoc current-state :available-service-ids available-service-ids'))))
+                (catch Exception e
+                  (log/error e "error in interstitial-maintainer")
+                  current-state))]
+          (when next-state
+            (recur next-state)))))
+    {:exit-chan exit-chan
+     :query-chan query-chan}))
+
+(let [interstitial-template-fn (template/fn
+                                 [{:keys [service-description service-id target-url]}]
+                                 (slurp (io/resource "web/interstitial.html")))]
+  (defn render-interstitial-template
+    "Renders the interstitial html page."
+    [context]
+    (interstitial-template-fn context)))
+
+(defn wrap-interstitial
+  "Redirects users to the interstitial page when interstitial has been enabled and the service has not been started.
+   The interstitial page will retry the request as a GET with the same query parameters but bypass the interstitial page."
+  [handler interstitial-state-atom]
+  (let [bypass-interstitial-param-name-value "x-waiter-bypass-interstitial=1"
+        text-html "text/html"]
+    (fn wrap-interstitial-handler [{:keys [descriptor headers query-string uri] :as request}]
+      (let [{:strs [accept host]} headers
+            {:keys [service-description service-id]} descriptor
+            {:strs [interstitial-secs]} service-description
+            ;; the bypass interstitial should be the last query parameter
+            bypass-interstitial? (str/ends-with? (str query-string) bypass-interstitial-param-name-value)]
+        (if (or bypass-interstitial? ;; bypass query parameter provided
+                (zero? interstitial-secs) ;; interstitial support has been disabled
+                (not (str/includes? (str accept) text-html)) ;; expecting html response
+                (not (interstitial-state-initialized? interstitial-state-atom))
+                (let [interstitial-promise (->> (partial process-interstitial-secs service-id interstitial-secs)
+                                                (start-service-interstitial! interstitial-state-atom service-id))]
+                  (or (nil? interstitial-promise)
+                      (realized? interstitial-promise))))
+          (->> (cond-> query-string
```

Comment:
We're relying on the followup bypass request to actually start the service, correct?